src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,210 @@
+            -1,
+            1));
+  }
+
+  public void holdPosition() {
+    setTurretPos(turretPosition);
+  }
+
+  public void zeroTurretEncoder() {
+    m_encoder.setPosition(0.0);
+  }
+
+  public void updateTurretPosition() {
+    turretPosition =
+        Math.toRadians(
+            ((m_encoder.getPosition().getValueAsDouble()) * 360)
+                * Constants.TurretConstants.CONVERSION_FACTOR);
+    SmartDashboard.putNumber("Turret Position", turretPosition);
+    SmartDashboard.putBoolean("Out of Bounds", Math.abs(turretPosition) > 3.14);
+    SmartDashboard.putBoolean(
+        "Soft limit enabled forward", turretMotor.isSoftLimitEnabled(SoftLimitDirection.kForward));
+    SmartDashboard.putBoolean(
+        "Soft limit enabled reverse", turretMotor.isSoftLimitEnabled(SoftLimitDirection.kReverse));
+  }
+
+  public void lockOntoTag(double rotationFromTag) {
+    turretMotor.set(turretPIDController.calculate(rotationFromTag, 0.0));
+  }
+
+  public double getTurretPos() {
+    return turretPosition; // returns the absolute encoder position in radians
+  }
+
+  public double getTurretVel() {
+    return turretVelocity;
+  }
+
+  public Command testTurretCommand() {
+    return runEnd(() -> setTurretPos(Math.toRadians(tunableTestPosition.get())), this::stopTurret);
+  }
+
+  public void LockTurret() {
+    lockOntoTag(Math.toRadians(LimelightHelpers.getTX("")));
+  }
+
+  public void TurretOdomCommand(Pose2d robotPos, Pose2d targetPos) {
+    setTurretPos(getTurretRotationFromOdometry(robotPos, targetPos));
+  }
+
+  public Command buildTurretCommand(boolean limelightVisible, Pose2d robotPos, Pose2d targetPos) {
+    return limelightVisible
+        ? run(this::LockTurret)
+        : runEnd(() -> TurretOdomCommand(robotPos, targetPos), this::stopTurret);
+  }
+
+  @Override
+  public void periodic() {
+    updateTurretPosition();
+    turretVelocity =
+        (m_encoder.getVelocity().getValueAsDouble())
+            * 60; // changing from rotations per second to rotations per minute or rpm
+    turretPositionEntry.log(turretPosition);
+    turretVelocityEntry.log(turretVelocity);
+  }
+
+  @Override
+  public void simulationPeriodic() {
+    updateTurretPosition();
+    turretVelocity =
+        (m_encoder.getVelocity().getValueAsDouble())
+            * 60; // changing from rotations per second to rotations per minute or rpm
+    turretPositionEntry.log(turretPosition);
+    turretVelocityEntry.log(turretVelocity);
+    for (var i = 0; i < Robot.defaultPeriodSecs; i += CANSparkMaxSim.kPeriod) {
+      turretSim.setInput(turretMotor.get() * RobotController.getBatteryVoltage());
+      turretSim.update(CANSparkMaxSim.kPeriod);
+      turretMotor.update(
+          turretSim.getVelocityRadPerSec() * Constants.TurretConstants.CONVERSION_FACTOR);
+    }
+    turretAngleSim.setAngle(Math.toDegrees(turretSim.getAngleRads()));
+    SmartDashboard.putNumber("Turret Angle", Math.toDegrees(turretSim.getAngleRads()));
+  }
```

Comment:
Since we are using the PIDController class and not the SparkPIDController, remove the for loop.